QCP-N-QSCD 411 1(412 2) and 412 5

breynders-cb opened the pull request:
As previously mentioned in https://github.com/digicert/pkilint/issues/124, a first PR which extends pkilint with QCP-N-QSCD for 411-1 (with restrictions of 412-2) and 412-5. 

I did my best in maintaining the existing structure of everything, please let me know where you'd want changes and we'll use this PR to get everything aligned as much as possible.

Additionally:
- I did not add tests, what's the testing policy? I see there's a lot of integration tests and I'm wondering if you have some tool/project set up that can already easily generate all these certificates in the right (i.e. wrong) format. If not I'll start building something for ourselves to generate some test cases for these types (but will take a bit).
- For etsi the `finding_metadata.csv` seems to be empty, I tried to document all the sources (and changes) but is it on the roadmap to fill in that csv?

And some further questions inlined ⬇️

breynders-cb commented on pkilint/etsi/en_319_412_5.py:
```diff
@@ -332,17 +352,25 @@ def retrieve_qualified_statement_id(cls, node):
+
         elif certificate_type in etsi_constants.NON_EU_QWAC_TYPES:
-            allowances[en_319_412_5.id_etsi_qcs_QcCompliance] = Rfc2119Word.MUST
+            # PR Question: Is this from 415_5.qcs-4.2? Needs different classifier?
             allowances[en_319_412_5.id_etsi_qcs_QcCClegislation] = Rfc2119Word.MUST
```
I was wondering what the source was for this rule. I couldn't really find it other than the reference in 412-5 QCS 4.2. If so, would it need a different source in the  validation finding?

breynders-cb commented on pkilint/etsi/en_319_412_5.py:
```diff
@@ -332,17 +352,25 @@ def retrieve_qualified_statement_id(cls, node):
+            # Table 2: 4.2.3 (QWAC is Annex IV)
+            if certificate_type in etsi_constants.EU_QWAC_TYPES:
+                allowances[en_319_412_5.id_etsi_qcs_QcType] = Rfc2119Word.MUST
+                # PR Question: Table 2, 4.2.2 only defines MUST, is the MUST_NOT also from 412-5 somewhere?
+                allowances[en_319_412_5.id_etsi_qcs_QcSSCD] = Rfc2119Word.MUST_NOT
```
In the referenced 412-5 / 5 table I couldn't find the `MUST_NOT` case anywhere, did I overlook something?

srdavidson replied:
My take is that eIDAS defines a QCSD in the context of e-signatures and e-seals https://eur-lex.europa.eu/legal-content/EN/TXT/HTML/?uri=CELEX:32014R0910#:~:text=for%20automated%20processing.-,ANNEX%20II,-REQUIREMENTS%20FOR%20QUALIFIED so it's not relevant to a QWAC

breynders-cb replied:
Thanks, that's clear!

breynders-cb commented:
Moved to ready-for-review since the ETSI rules are now considered feature-complete from my end (and I'm going to shift towards implementing the POR rules now), looking forward to the feedback!

CBonnell commented:
Thank you this great contribution, @breynders-cb! I'm currently traveling for work this week, but will review this PR fully when I return next week.

As for the test case generation, we use [der-ascii](https://github.com/google/der-ascii) to generate test artifacts. It has a bit of learning curve to use, but quite powerful and flexible. The test case file format is the PEM text of the artifact followed by the CSV-formatted output of findings. This makes it relatively simple to write test case generation scripts.

We originally did not flesh out the ETSI `finding_metadata.csv` file, as the ETSI linter references the citation directly in the finding codes. We can certainly flesh out that file if that would be helpful. We are planning to revamp and improve the documentation for findings as part of #64, but haven't had the cycles yet do to so.

breynders-cb commented:
> Thank you this great contribution, @breynders-cb! I'm currently traveling for work this week, but will review this PR fully when I return next week.
> 
> As for the test case generation, we use [der-ascii](https://github.com/google/der-ascii) to generate test artifacts. It has a bit of learning curve to use, but quite powerful and flexible. The test case file format is the PEM text of the artifact followed by the CSV-formatted output of findings. This makes it relatively simple to write test case generation scripts.
> 
> We originally did not flesh out the ETSI `finding_metadata.csv` file, as the ETSI linter references the citation directly in the finding codes. We can certainly flesh out that file if that would be helpful. We are planning to revamp and improve the documentation for findings as part of #64, but haven't had the cycles yet do to so.

Great, thanks! I'll add der-ascii to my list of tools, for now I spruced up some of our test code and generated test certificates through bouncy castle so all new rules should have tests in the PEM+csv format.

CBonnell commented on pkilint/etsi/etsi_shared.py:
```diff
@@ -147,3 +149,111 @@ def validate(self, node):
+
+
+class LegalPersonDuplicateAttributeAllowanceValidator(validation.Validator):
+    """
```
These citation comments should probably be moved to the appropriate sub-class so they document the validation finding declaration.

CBonnell commented on pkilint/etsi/en_319_412_5.py:
```diff
@@ -307,6 +309,24 @@ def __init__(self):
         )
 
 
+class QcStatementPresenceValidator(extension.ExtensionPresenceValidator):
```
I don't think this class is needed, as the `NaturalPersonExtensionIdentifierAllowanceValidator` flags when the QCStatements extension is missing.

breynders-cb replied:
I might be wrong but I don't think the NaturalPersonExtensionIdentifierAllowanceValidator checks presence of the QCStatements if they're completely missing. 

In https://github.com/cleverbase/pkilint/blob/qcp-n-qscd-411-1(412-2)-and-412-5/tests/integration_certificate/etsi/qcp_n_qscd_final_certificate/missing_qc_statements.crttest I tried to specifically test this but maybe I messed that one up somehow?

CBonnell reviewed (commented):
Thanks for this contribution! Overall, I think this PR is in pretty good shape. I added some comments for relatively minor things.

In regard to the development of test cases, do you plan on adding more, or should I assist in creating them?

breynders-cb commented:
> Thanks for this contribution! Overall, I think this PR is in pretty good shape. I added some comments for relatively minor things.
> 
> In regard to the development of test cases, do you plan on adding more, or should I assist in creating them?

For test cases I worked mostly with integration tests by generating certificates so I'd like to leave it at that for this PR. I'll go through your comments and address them ASAP, thanks for the review!

breynders-cb commented:
Rebased and addressed all comments, could you have a look again? I had 1 question left w.r.t. the `QcStatementPresenceValidator`, agreed on all other comments.

CBonnell commented:
Thanks @breynders-cb for the fast turnaround, I'll review your changes this week.

CBonnell commented on tests/integration_certificate/etsi/qncp_w_gen_legal_person_eidas_final_certificate/rsapss_sig_alg.crttest:
```diff
@@ -34,3 +34,4 @@ certificate.tbsCertificate.extensions.4.extnValue.subjectKeyIdentifier,SubjectKe
 certificate.tbsCertificate.extensions,NaturalPersonExtensionIdentifierAllowanceValidator,ERROR,etsi.en_319_412_2.qc_statements_extension_absent,
 certificate.tbsCertificate.subject.rdnSequence.3.0.value.x520CommonName,NcpWCommonNameValidator,ERROR,etsi.en_319_412_4.web-4.1.3-4.common_name_unknown_source,"Unknown source for value of common name: ""sct"""
 certificate.tbsCertificate.extensions,NcpWSubjectAltNamePresenceValidator,ERROR,etsi.en_319_412_4.web-4.1.3-4.san_missing,
+certificate.tbsCertificate.extensions,QcStatementPresenceValidator,ERROR,etsi.en_319_412_5.qcs-5.01,
```
This finding is duplicative of the one on line 34, but I think this will go away once the natural person extension validator is amended to check QCP-N-QSCD/`EU_TYPES`.

CBonnell reviewed (commented):
This looks great, thank you! I added a few comments, mainly surrounding the potentially duplicative QCStatements extension check (one of the existing classes can be simply amended to require QCStatements for all `EU_TYPES`).

CBonnell commented:
> Addressed all comments, could you have a look again?

Thanks for the updates, will take a look in the next day or two.
